src/osd_field.c

```diff
@@ -0,0 +1,359 @@
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_output_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: o */
+	if (wl_list_length(&view->server->outputs) > 1 &&
+			output_is_usable(view->output)) {
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_identifier(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: I */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ false));
+}
+
+static void
+field_set_identifier_trimmed(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: i */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ true));
+}
+
+static void
+field_set_title(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: T */
+	buf_add(buf, get_title(view));
+}
+
+static void
+field_set_title_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: t */
+	buf_add(buf, get_title_if_different(view));
+}
+
+static void
+field_set_custom(struct buf *buf, struct view *view, const char *format)
+{
+	if (!format) {
+		wlr_log(WLR_ERROR, "Missing format for custom window switcher field");
+		return;
+	}
+
+	char fmt[LAB_FIELD_SINGLE_FMT_MAX_LEN];
+	unsigned char fmt_len = 0;
+
+	struct buf field_result;
+	buf_init(&field_result);
+
+	char field_formatted[4096];
+
+	for (const char *p = format; *p; p++) {
+		if (!fmt_len) {
+			if (*p == '%') {
+				fmt[fmt_len++] = *p;
+			} else {
+				/*
+				 * Just relay anything not part of a
+				 * format string to the output buffer.
+				 */
+				buf_add_char(buf, *p);
+			}
+			continue;
+		}
+
+		/* Allow string formatting */
+		/* TODO: add . for manual truncating? */
+		if (*p == '-' || *p == '#' || (*p >= '0' && *p <= '9')) {
+			if (fmt_len >= LAB_FIELD_SINGLE_FMT_MAX_LEN - 2) {
+				/* Leave space for terminating 's' and NULL byte */
+				wlr_log(WLR_ERROR,
+					"single format string length exceeded: '%s'", p);
+			} else {
+				fmt[fmt_len++] = *p;
+			}
+			continue;
+		}
+
+		/* Handlers */
+		for (unsigned char i = 0; i < LAB_FIELD_COUNT; i++) {
+			if (*p != field_handlers[i].fmt_char) {
+				continue;
+			}
+
+			/* Generate the actual content*/
+			field_handlers[i].fn(&field_result, view, /*format*/ NULL);
+
+			/* Throw it at snprintf to allow formatting / padding */
+			fmt[fmt_len++] = 's';
+			fmt[fmt_len++] = '\0';
+			snprintf(field_formatted, sizeof(field_formatted),
+				fmt, field_result.buf);
+
+			/* And finally write it to the output buffer */
+			buf_add(buf, field_formatted);
+			goto reset_format;
+		}
+
+		wlr_log(WLR_ERROR,
+			"invalid format character found for osd %s: '%c'",
+			format, *p);
+
+reset_format:
+		/* Reset format string and tmp field result buffer */
+		buf_clear(&field_result);
+		fmt_len = 0;
+	}
+	free(field_result.buf);
+}
+
+static const struct field_handlers field_handlers[LAB_FIELD_COUNT] = {
```

Comment:
I'm okay with this, but my preferred style is 
```
static struct foo foo[] { {
  .content = WHATEVER,
  .conversion_specifier = 'B',
  .convert = some_function_pointer
}, {
...etc..
};
```

...and then use `ARRAY_SIZE()`.

Reply:
This array is accessed directly via the enum as index as well (via `osd_field_get_content()` from `osd.c`).

Reply:
Yes, good point. Happy with approach.